File: src/main/java/eu/happybe/openapi/form/FormQueue.java
package eu.happybe.openapi.form;

import cn.nukkit.Player;
import cn.nukkit.network.protocol.ModalFormRequestPacket;
import cn.nukkit.network.protocol.ModalFormResponsePacket;
import eu.happybe.openapi.OpenAPI;

import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

public class FormQueue {

    public static Map<String, Map<Integer, Form>> formQueue = new HashMap<>();

    public static void sendForm(Player player, Form form) {
        int formId = FormQueue.getWindowId(player);
        if(!FormQueue.formQueue.containsKey(player.getName())) {
            FormQueue.formQueue.put(player.getName(), new HashMap<>());
        }

        FormQueue.formQueue.get(player.getName()).put(formId, form);

        ModalFormRequestPacket pk = new ModalFormRequestPacket();
        pk.formId = formId;
        pk.data = form.serialize();

        player.dataPacket(pk);
    }

    private static int getWindowId(Player player) {
        int windowId;

        try {
            Field field = player.getClass().getDeclaredField("formWindowCount");
            field.setAccessible(true);

            windowId = field.getInt(player);
            field.setInt(player,windowId + 1);
        }
        catch (NoSuchFieldException | IllegalAccessException e) {
            OpenAPI.getInstance().getLogger().error(e.getMessage());
            e.printStackTrace();

            windowId = new Random().nextInt();
        }

        return windowId;
    }

    public static void handleQuit(Player player) {
        FormQueue.formQueue.remove(player.getName());
    }

    public static boolean handleFormDataPacket(Player player, ModalFormResponsePacket packet) {
        if(!FormQueue.formQueue.containsKey(player.getName())) {
            return false;
        }
        if(!FormQueue.formQueue.get(player.getName()).containsKey(packet.formId)) {
            return false;
        }

        Form form = FormQueue.formQueue.get(player.getName()).get(packet.formId);

        if(form.getCallback() != null && !packet.data.trim().equals("null")) { // Currently we aren't able to handle closing, maybe I will implement it later, but we don't use this as a functionality
            form.getCallback().accept(new FormCallbackResponse(player, form, packet.data));
        }

        FormQueue.formQueue.get(player.getName()).remove(packet.formId);
        return true;
    }
}

File: src/main/java/eu/happybe/openapi/form/FormQueue.kt
package eu.happybe.openapi.form

import cn.nukkit.Player
import cn.nukkit.network.protocol.ModalFormRequestPacket
import cn.nukkit.network.protocol.ModalFormResponsePacket
import eu.happybe.openapi.OpenAPI
import java.util.*

object FormQueue {
    var formQueue: MutableMap<String, MutableMap<Int, Form>> = HashMap()
    fun sendForm(player: Player, form: Form) {
        val formId = getWindowId(player)
        if (!formQueue.containsKey(player.name)) {
            formQueue[player.name] = HashMap()
        }
        formQueue[player.name]!![formId] = form
        val pk = ModalFormRequestPacket()
        pk.formId = formId
        pk.data = form.serialize()
        player.dataPacket(pk)
    }

    private fun getWindowId(player: Player): Int {
        var windowId: Int
        try {
            val field = player.javaClass.getDeclaredField("formWindowCount")
            field.isAccessible = true
            windowId = field.getInt(player)
            field.setInt(player, windowId + 1)
        } catch (e: NoSuchFieldException) {
            OpenAPI.getInstance().logger.error(e.message)
            e.printStackTrace()
            windowId = Random().nextInt()
        } catch (e: IllegalAccessException) {
            OpenAPI.getInstance().logger.error(e.message)
            e.printStackTrace()
            windowId = Random().nextInt()
        }
        return windowId
    }

    fun handleQuit(player: Player) {
        formQueue.remove(player.name)
    }

    fun handleFormDataPacket(player: Player, packet: ModalFormResponsePacket): Boolean {
        if (!formQueue.containsKey(player.name)) {
            return false
        }
        if (!formQueue[player.name]!!.containsKey(packet.formId)) {
            return false
        }
        val form = formQueue[player.name]!![packet.formId]
        if (form.getCallback() != null && packet.data.trim { it <= ' ' } != "null") { // Currently we aren't able to handle closing, maybe I will implement it later, but we don't use this as a functionality
            form.getCallback().accept(FormCallbackResponse(player, form, packet.data))
        }
        formQueue[player.name]!!.remove(packet.formId)
        return true
    }
}